File: hexameter-core/src/test/java/org/codetome/hexameter/core/internal/GridDataTest.java
package org.codetome.hexameter.core.internal;

import org.codetome.hexameter.core.api.HexagonOrientation;
import org.codetome.hexameter.core.api.HexagonalGridLayout;
import org.junit.After;
import org.junit.Test;

import static java.lang.Math.sqrt;
import static junit.framework.Assert.assertEquals;
import static org.codetome.hexameter.core.api.HexagonOrientation.FLAT_TOP;
import static org.codetome.hexameter.core.api.HexagonOrientation.POINTY_TOP;
import static org.codetome.hexameter.core.api.HexagonalGridLayout.RECTANGULAR;

public class GridDataTest {

    private static final HexagonOrientation ORIENTATION = FLAT_TOP;
    private static final HexagonalGridLayout GRID_LAYOUT = RECTANGULAR;
    private static final double RADIUS = 30;
    private static final int GRID_WIDTH = 30;
    private static final int GRID_HEIGHT = 30;

    private GridData target;

    @After
    public void tearDown() {
        target = null;
    }

    @Test
    public void shouldProperlyReturnRadiusWhenGetRadiusIsCalled() {
        target = new GridData(ORIENTATION, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT);
        assertEquals(RADIUS, target.getRadius());
    }

    @Test
    public void shouldProperlyCalculateWidthWithPointyHexagonsWhenGetWidthIsCalled() {
        target = createWithPointy();
        final double expectedWidth = sqrt(3) * RADIUS;
        final double actualWidth = target.getHexagonWidth();
        assertEquals(expectedWidth, actualWidth);
    }

    private GridData createWithPointy() {
        return new GridData(POINTY_TOP, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT);
    }

    @Test
    public void shouldProperlyCalculateWidthWithFlatHexagonsWhenGetWidthIsCalled() {
        target = createWithFlat();
        final double expectedWidth = RADIUS * 3 / 2;
        final double actualWidth = target.getHexagonWidth();
        assertEquals(expectedWidth, actualWidth);
    }

    private GridData createWithFlat() {
        return new GridData(FLAT_TOP, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT);
    }

    @Test
    public void shouldProperlyCalculateHeightWithPointyHexagonsWhenGetHeightIsCalled() {
        target = createWithPointy();
        final double expectedHeight = RADIUS * 3 / 2;
        final double actualHeight = target.getHexagonHeight();
        assertEquals(expectedHeight, actualHeight);
    }

    @Test
    public void shouldProperlyCalculateHeightWithFlatHexagonsWhenGetHeightIsCalled() {
        target = createWithFlat();
        final double expectedHeight = sqrt(3) * RADIUS;
        final double actualHeight = target.getHexagonHeight();
        assertEquals(expectedHeight, actualHeight);
    }

    @Test
    public void shouldReturnProperOrientationWhenGetOrientationIsCalled() {
        target = new GridData(ORIENTATION, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT);
        assertEquals(ORIENTATION, target.getOrientation());
    }

    @Test
    public void shouldReturnProperCoordinateOffsetWhengetCoordinateOffsetIsCalled() {
        target = new GridData(ORIENTATION, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT);
        assertEquals(ORIENTATION.getCoordinateOffset(), target.getOrientation().getCoordinateOffset());
    }
}

File: mixite2.core/src/commonTest/kotlin/org/hexworks/mixite2/core/internal/GridDataTest.kt
package org.hexworks.mixite2.core.internal

import org.hexworks.mixite2.core.api.HexagonOrientation.FLAT_TOP
import org.hexworks.mixite2.core.api.HexagonOrientation.POINTY_TOP
import org.hexworks.mixite2.core.api.HexagonalGridLayout.RECTANGULAR
import kotlin.math.sqrt
import kotlin.test.Test
import kotlin.test.assertEquals

class GridDataTest {

    lateinit var target: GridData

    @Test
    fun shouldProperlyReturnRadiusWhenGetRadiusIsCalled() {
        target = GridData(ORIENTATION, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT)
        assertEquals(RADIUS, target.radius)
    }

    @Test
    fun shouldProperlyCalculateWidthWithPointyHexagonsWhenGetWidthIsCalled() {
        target = createWithPointy()
        val expectedWidth = sqrt(3.0) * RADIUS
        val actualWidth = target.hexagonWidth
        assertEquals(expectedWidth, actualWidth)
    }

    private fun createWithPointy(): GridData {
        return GridData(POINTY_TOP, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT)
    }

    @Test
    fun shouldProperlyCalculateWidthWithFlatHexagonsWhenGetWidthIsCalled() {
        target = createWithFlat()
        val expectedWidth = RADIUS * 3 / 2
        val actualWidth = target.hexagonWidth
        assertEquals(expectedWidth, actualWidth)
    }

    private fun createWithFlat(): GridData {
        return GridData(FLAT_TOP, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT)
    }

    @Test
    fun shouldProperlyCalculateHeightWithPointyHexagonsWhenGetHeightIsCalled() {
        target = createWithPointy()
        val expectedHeight = RADIUS * 3 / 2
        val actualHeight = target.hexagonHeight
        assertEquals(expectedHeight, actualHeight)
    }

    @Test
    fun shouldProperlyCalculateHeightWithFlatHexagonsWhenGetHeightIsCalled() {
        target = createWithFlat()
        val expectedHeight = sqrt(3.0) * RADIUS
        val actualHeight = target.hexagonHeight
        assertEquals(expectedHeight, actualHeight)
    }

    @Test
    fun shouldReturnProperOrientationWhenGetOrientationIsCalled() {
        target = GridData(ORIENTATION, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT)
        assertEquals(ORIENTATION, target.orientation)
    }

    @Test
    fun shouldReturnProperCoordinateOffsetWhengetCoordinateOffsetIsCalled() {
        target = GridData(ORIENTATION, GRID_LAYOUT, RADIUS, GRID_WIDTH, GRID_HEIGHT)
        assertEquals(ORIENTATION.coordinateOffset, target.orientation.coordinateOffset)
    }

    companion object {

        private const val RADIUS = 30.0
        private const val GRID_WIDTH = 30
        private const val GRID_HEIGHT = 30
        private val ORIENTATION = FLAT_TOP
        private val GRID_LAYOUT = RECTANGULAR.gridLayoutStrategy
    }
}
